src/watch_state.rs

```diff
@@ -0,0 +1,231 @@
+        tracing::debug!("handling effect {:?} in current state {:?}", effect, *state);
+        let outcome = match (effect, *state) {
+            (Effect::Exit, _) => {
+                Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+            }
+            (Effect::ProcessFailed, _) => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+            (Effect::ProcessCompleted, State::Building) => {
+                *state = State::Running;
+                self.restart_outcome(true)
+            }
+            (Effect::ProcessCompleted, State::Running) => Outcome::DoNothing,
+            (Effect::ManifestChange, State::Building) => self.restart_outcome(false),
+            (Effect::ManifestChange, State::Running) => {
+                if !self.skip_build {
+                    *state = State::Building;
+                }
+                self.restart_outcome(false)
+            }
+            (Effect::SourceChange, State::Building) => self.restart_outcome(false),
+            (Effect::SourceChange, State::Running) => {
+                if !self.skip_build {
+                    *state = State::Building;
+                    return self.restart_outcome(false);
+                }
+                Outcome::DoNothing
+            }
+            (Effect::ArtifactChange, State::Building) => Outcome::DoNothing,
+            (Effect::ArtifactChange, State::Running) => self.restart_outcome(false),
+            (Effect::DoNothing, _) => Outcome::if_running(Outcome::DoNothing, Outcome::Start),
+        };
+        tracing::debug!("now in {:?} state with outcome {:?}", *state, outcome);
+        outcome
+    }
+
+    fn restart_outcome(&self, prevent_clear: bool) -> Outcome {
+        let should_clear = !prevent_clear && self.clear;
+        Outcome::sequence(
+            [
+                Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+                match should_clear {
+                    true => Outcome::Clear,
+                    false => Outcome::DoNothing,
+                },
+                Outcome::Start,
+            ]
+            .into_iter(),
+        )
+    }
+}
+
+/// A state that the state machine can be in.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum State {
+    Building,
+    Running,
+}
+
+/// An effect is parsed from the events of an action and results in an outcome.
+///
+/// The variants are ordered by highest to lowest precedence so that they can be sorted. When an
+/// action has multiple events (this occurs when events are debounced) they will all produce
+/// effects. The highest precedence effect will be chosen to produce the outcome.
+#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub enum Effect {
+    /// Exit spin watch
+    Exit,
+    /// Either `spin build` or `spin up` failed to run
+    ProcessFailed,
+    /// `spin build` has completed running, `spin up` never completes
+    ProcessCompleted,
```

Comment:
Reading where this was consumed I wasn't sure what "process" it referred to, whether some internal procedure or an actual system process.  Maybe rename to `BuildCompleted` or `ChildProcessCompleted`?

Reply:
I like it.